nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java

```diff
@@ -0,0 +1,293 @@
+//            - Milestone 2 or 3, not Milestone 1.
+//      - Remove test-pypi usage from ExtensionManager.py
+//      - Additional Interfaces beyond just FlowFileTransform
+//          - FlowFileSource
+//      - Restructure Maven projects
+//          - Should this all go under Framework?
+//
+//
+//      CONSIDER:
+//      - Clustering: Ensure component on all nodes?
+//          - Consider "pip freeze" type of thing to ensure that python dependencies are same across nodes when joining cluster.
+//      - Update python code to use python_style_method_names instead of javaStyleMethodNames
+//      - Also add 'failure' and 'original' relationships to FlowFileTransform
+//
+//
+//      Can punt for now:
+//      - We have an issue with objects created from Processor calling into Java. Is fine when we provide objects to Python but when
+//          it makes a callback, those objects are bound and never unbound!!!
+//              *** This appears to be fine as long as on the Python side we set manage_memory=True ***
+//                  But it does cut perf in half. May be room for improvement somehow? By implementing a 'bulk delete' custom command?
+//      - Appears to still be issues with timeout in nifi when set to 10 secs... initially will probably just leave default of 0sec.
+
+public class PythonProcess {
+    private static final Logger logger = LoggerFactory.getLogger(PythonProcess.class);
+    private static final String PYTHON_CONTROLLER_FILENAME = "Controller.py";
+
+    private final PythonProcessConfig processConfig;
+    private final ControllerServiceTypeLookup controllerServiceTypeLookup;
+    private final File virtualEnvHome;
+    private GatewayServer server;
+    private PythonController controller;
+    private Process process;
+    private NiFiPythonGateway gateway;
+    private final Map<String, Boolean> processorPrefersIsolation = new ConcurrentHashMap<>();
+
+
+    public PythonProcess(final PythonProcessConfig processConfig, final ControllerServiceTypeLookup controllerServiceTypeLookup, final File virtualEnvHome) {
+        this.processConfig = processConfig;
+        this.controllerServiceTypeLookup = controllerServiceTypeLookup;
+        this.virtualEnvHome = virtualEnvHome;
+    }
+
+    public PythonController getController() {
+        return controller;
+    }
+
+    public void start() throws IOException {
+        // TODO: Look into using configured TLS Certs to make this secure by default.
+        final ServerSocketFactory serverSocketFactory = ServerSocketFactory.getDefault();
+        final SocketFactory socketFactory = SocketFactory.getDefault();
+
+        final int timeoutMillis = (int) processConfig.getCommsTimeout().toMillis();
+        final String authToken = null;
+        final CallbackClient callbackClient = new CallbackClient(GatewayServer.DEFAULT_PYTHON_PORT, GatewayServer.defaultAddress(), authToken,
+            50000L, TimeUnit.MILLISECONDS, socketFactory, false, timeoutMillis);
```

Comment:
I would actually disagree with this. Naming the variable here makes it more clear when reading from GitHub. But in a typical IDE such as IntelliJ it already knows the argument name and makes it very clear. Introducing a named constant then means that we're making the code more difficult to read by having to jump to a different part of the code to see the value. So while introducing the named constant helps in one case and hurts in another, we should optimize for the case that's most common. I'd say that case, by far, is IntelliJ. Of course, if the value were referenced multiple times, etc. then it would make sense to name it but in this case I prefer the raw constant value.

Reply:
Thanks for the thoughtful reply, good point about IDE view versus GitHub view. Leaving it as-is sounds good.